File: QuynhLuu_2324/Bai3/Sortx.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>
#include <algorithm>

bool isNumeric(const std::string& str) {
    return std::all_of(str.begin(), str.end(), ::isdigit);
}

int main() {
    std::ifstream infile("Sortx.INP");
    std::ofstream outfile("Sortx.OUT");

    int n;
    infile >> n;
    infile.ignore(); // Bỏ qua dòng mới sau khi đọc n

    std::vector<std::string> strings(n);
    std::string line;
    std::getline(infile, line);
    std::istringstream iss(line);

    // Đọc vào tất cả các chuỗi
    for (int i = 0; i < n; ++i) {
        iss >> strings[i];
    }

    // Tách các chuỗi số và không phải số
    std::vector<std::string> numeric_strings;
    std::vector<std::string> non_numeric_strings;

    for (const auto& str : strings) {
        if (isNumeric(str)) {
            numeric_strings.push_back(str);
        } else {
            non_numeric_strings.push_back(str);
        }
    }

    // Sắp xếp các chuỗi số
    std::sort(numeric_strings.begin(), numeric_strings.end(), [](const std::string& a, const std::string& b) {
        return std::stoi(a) < std::stoi(b);
    });

    // Kết hợp lại danh sách theo thứ tự ban đầu
    std::vector<std::string> result;
    auto num_it = numeric_strings.begin();
    for (const auto& str : strings) {
        if (isNumeric(str)) {
            result.push_back(*num_it);
            ++num_it;
        } else {
            result.push_back(str);
        }
    }

    // Ghi kết quả ra file
    for (const auto& str : result) {
        outfile << str << " ";
    }
    outfile << std::endl;

    infile.close();
    outfile.close();

    return 0;
}

File: THCS/2023_2024/QuynhLuu_2324/Bai3/Sortx.py
def sort_strings(strings):
    # Tách các chuỗi số và chuỗi không phải số
    numeric_strings = [s for s in strings if s.isdigit()]
    non_numeric_strings = [s for s in strings if not s.isdigit()]
    
    # Sắp xếp các chuỗi số
    numeric_strings.sort(key=int)

    # Kết hợp lại danh sách theo thứ tự ban đầu
    result = []
    num_index = 0
    for s in strings:
        if s.isdigit():
            result.append(numeric_strings[num_index])
            num_index += 1
        else:
            result.append(s)
    
    return result

def main():
    # Đọc dữ liệu từ file Sortx.INP
    with open("Sortx.INP", "r") as infile:
        n = int(infile.readline().strip())  # Số lượng chuỗi
        strings = infile.readline().strip().split()  # Danh sách các chuỗi

    # Sắp xếp các chuỗi
    sorted_strings = sort_strings(strings)

    # Ghi kết quả ra file Sortx.OUT
    with open("Sortx.OUT", "w") as outfile:
        outfile.write(" ".join(sorted_strings) + "\n")

if __name__ == "__main__":
    main()
